[SLO] Introduce Custom Metric SLI

simianhacker opened the pull request:
## Summary

This PR introduces a Custom Metric SLI for SLOs. This will allow the user to select multiple fields to aggregate with a `sum` aggregation and use them in an equation to express "good" events and "total" events.  The equation field can do simple math like `(A + B) / C` along with simple boolean logic like `A > B ? A : B`. The valid operators for are: `+, -, /, *, (, ), ?, !, &, :, |, >, <, =`.

![image](https://github.com/elastic/kibana/assets/41702/43e6fd09-b359-46fc-8b4d-5de0301e09b2)

![image](https://github.com/elastic/kibana/assets/41702/b9603af7-2814-42fe-9910-2794e969bb92)

simianhacker committed:
[SLO] Introduce custom metric SLI

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

simianhacker committed:
Fixing types

simianhacker committed:
Adding form validation for the equation

simianhacker committed:
Fixing slo_edit test

simianhacker committed:
Changing metric.id to metric.name to avoid conflicts with react-form-hook; adding form validation

simianhacker committed:
Reusing MetricCustomIndicatorSchema

elasticmachine commented:
Pinging @elastic/actionable-observability (Team: Actionable Observability)

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into custom-metric-sli

CoenWarmer commented:
Super cool stuff, nice to see this new SLI added so quickly.

Just some feedback on the UI, (and apologies if these are all dumb remarks):

<img width="762" alt="Screenshot 2023-05-15 at 16 37 40" src="https://github.com/elastic/kibana/assets/535564/e44da060-d17d-4565-9f55-c9c61b638b37">

I was a little confused by the "Sum of" prepend part of the inputs. I initially thought they were clickable and would provide options such as divide by / product of, etc, but it seems they are always "Sum of".  This makes sense, but this text in combination with the equation field below threw me off. The fields that are selected are always "Sum of", right? And with the equation I would be able to subtract, divide etc? 

Other thing: would it be a good idea to have a EuiHelpText next to the "Good equation" and "Total equation" form labels that give an example of an acceptable equation?

And a very minor thing is I would add just a little more vertical spacing between the "Add metric" button and "Good equation" label.

CoenWarmer reviewed (commented):
Looks great already. Some small comments.

simianhacker commented on x-pack/plugins/observability/public/pages/slo_edit/components/slo_edit_form_indicator_section.tsx:
```diff
@@ -8,17 +8,38 @@
+      setValue('indicator.params.total.metrics', [NEW_CUSTOM_METRIC]);
+      setValue('indicator.params.total.equation', 'A');
+    }
+    if (indicator === 'sli.kql.custom' && isObject(watch('indicator.params.good'))) {
```
I needed to do this for the `sli.kql.custom` and `sli.metric.custom` because both indicators use `good` and `total` as part of their params. For the `sli.metric.custom` I needed to set the default values.

simianhacker committed:
Removing APM type casts; keeping KQL cast because Typescript is tripping

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into custom-metric-sli

simianhacker committed:
Adding space below 'Add metric' button

simianhacker committed:
Adding additional help text

simianhacker committed:
fixing typo

simianhacker commented:
> I was a little confused by the "Sum of" prepend part of the inputs. I initially thought they were clickable and would provide options such as divide by / product of, etc, but it seems they are always "Sum of". This makes sense, but this text in combination with the equation field below threw me off. The fields that a user selects are always "Sum of", right? And with the equation I am able to subtract, divide etc? I'm wondering if the "Sum of" could be removed then as that is implied?

@CoenWarmer I wanted to make sure the user understood that we are running a "sum aggregation" on these fields.

@katrin-freihofner @grabowskit What's your thoughts on keeping `Sum of` prefix on these fields?

> Other thing: would it be a good idea to have a EuiHelpText next to the "Good equation" and "Total equation" form labels that give an example of an acceptable equation?

Good idea,  I added some help text next to the labels and expanded the text under the equation field to include all the valid characters.

<img width="817" alt="image" src="https://github.com/elastic/kibana/assets/41702/2606bb0f-c88e-4cc5-b371-807afa12421d">

<img width="823" alt="image" src="https://github.com/elastic/kibana/assets/41702/74b1c614-0d7e-4f13-b16f-b4964fe03635">



> And a very minor thing is I would add just a little more vertical spacing between the "Add metric" button and "Good equation" label.

Fixed

kdelemme reviewed (commented):
Thanks for building this! I did a first pass, I'm going to test it locally now.

simianhacker committed:
Update x-pack/plugins/observability/dev_docs/slo.md

Co-authored-by: Kevin Delemme <kdelemme@gmail.com>

simianhacker commented on x-pack/plugins/observability/public/pages/slo_edit/components/slo_edit_form_indicator_section.tsx:
```diff
@@ -8,17 +8,38 @@
+
+  // Set the defaults for the form when switching between different indicators
+  useEffect(() => {
+    if (indicator === 'sli.metric.custom') {
+      setValue('indicator.params.good.metrics', [NEW_CUSTOM_METRIC]);
+      setValue('indicator.params.good.equation', 'A');
+      setValue('indicator.params.total.metrics', [NEW_CUSTOM_METRIC]);
+      setValue('indicator.params.total.equation', 'A');
+    }
```
Yes... if you started on "Custom metric" then switched to "Custom KQL" then back, you would get a fresh form. The reason being is that `indicator.params.good` (and `total`) is a different shaped object on both on `Custom KQL`  (`string`) and `Custom Metric` (`object`). If we wanted to keep the state when switching between indicators, I would need to store `indicator.params.good` in something like `indicator.params.goodMetric`. 

This begs the question, "is it reasonable to expect the form to retain state when switching the indicator types?"

kdelemme replied:
I understand the reason but I'm wondering in the case of the **edit** flow (i'm gonna test this AM so I'll check that), if it does not erase the good and total metrics/equation when it loads the form?

> This begs the question, "is it reasonable to expect the form to retain state when switching the indicator types?"

I think for the sake of code simplicity and maintenance, that would be a fine approach by me.

kdelemme replied:
> I understand the reason but I'm wondering in the case of the edit flow (i'm gonna test this AM so I'll check that), if it does not erase the good and total metrics/equation when it loads the form?

tested and it does not, all good.

But I think we should maybe reset the form when we switch indicator type. It will simplify everything. wdyt?

simianhacker committed:
Updating dev_docs to match new format

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into custom-metric-sli

simianhacker commented on x-pack/plugins/observability/docs/openapi/slo/components/schemas/indicator_properties_custom_metric.yaml:
```diff
@@ -0,0 +1,106 @@
+        example: 'field.environment : "production" and service.name : "my-service"'
+      timestampField:
+        description: >
+          The timestamp field used in the source indice. If not specified, @timestamp will be used.
```
if you look on `line 14` you can see it's listed as a required field, same on the KQL:

https://github.com/elastic/kibana/blob/b9f05dffd02e1f8baef65a9f0d906595e55751cc/x-pack/plugins/observability/docs/openapi/slo/components/schemas/indicator_properties_custom_kql.yaml#L12-L14

simianhacker replied:
Ah... I see what you mean. I removed ` If not specified, @timestamp will be used.` from both Custom KQL and Custom Metric docs.

simianhacker committed:
Update OpenAPI docs to for timestampField to reflect reality

kdelemme commented:
### Manual testing

_work in progress_

- [ ] When **creating** or **editing** a custom metric SLI, then switching indicator to **Custom KQL** and then back to **Custom Metric**, you'll see an error:
![image](https://github.com/elastic/kibana/assets/1376800/5915f020-65b4-4cfd-bd65-3c5fe31d5132)

kdelemme reviewed (changes requested):
the section validation can lead to a "cannot read .. from undefined"

simianhacker committed:
Update x-pack/plugins/observability/public/pages/slo_edit/components/slo_edit_form_indicator_section.tsx

Co-authored-by: Kevin Delemme <kdelemme@gmail.com>

simianhacker committed:
Update x-pack/plugins/observability/public/pages/slo_edit/hooks/use_section_form_validation.ts

Co-authored-by: Kevin Delemme <kdelemme@gmail.com>

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

simianhacker committed:
Remove useWatch

simianhacker committed:
remove SLI selector; set the defaults when selecting which indictor to
display

simianhacker committed:
Adding isObject check

simianhacker committed:
Fixing the defaults for sli.kql.custom and sli.metric.custom; fixing editing

simianhacker committed:
hacking the hidden fields without the Control jsx

simianhacker committed:
Updating test to remove the SLI indictor

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into custom-metric-sli

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into custom-metric-sli

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into custom-metric-sli

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into custom-metric-sli

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into custom-metric-sli

simianhacker committed:
Replacing replace with append/remove

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into custom-metric-sli

simianhacker committed:
removing unused type

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into custom-metric-sli

kdelemme reviewed (approved):
LGTM!
